vivo-project/Vitro
# [VIVO-1545] Tracking of user changes to the content store

**[VIVO-1545](https://jira.duraspace.org/browse/VIVO-1545)**: (please link to issue)

# What does this pull request do?
The pull request adds a module that enables tracking of changes being made in the triple store by users through the Vitro user interface.

Changes are recorded in a triple store, with the users ID (URI), the time, and the changes that have been made.

A user interface (/audit) is also provided that, when logged in, displays the changes that have been made by that user.

# What's new?
Adds a new AuditModule application module
TDBAuditModule can be configured / enabled via the applicationSetup.n3
A servlet request listener and RDFService change listener work in tandem to capture any changes made to the content store, and record those changes in a dedicated TDB store.

# How should this be tested?
An unmodified applicationSetup.n3 should result in no changes to the application behaviour / home directory.
Adding the TDBAuditModule configuration to applicationSetup.n3 will:
* create a TDB store in the file system (the recommendation is to be "tdbAuditModels" within the Vitro home directory)
* Any modifications to the data (e.g. editing a Person's overview) will result in the changes being logged in audit TDB store, but otherwise there will be no difference to the perceived behaviour of the application.
* Whilst logged in, accessing /audit will show the changes that the logged in user has previously made

Everything else about the application should operate normally - every request will be going through the servlet request listener, but it will not be doing any work when there is no logged in user / changes made to the content store.

# Additional Notes:
As it is adding a listener to the RDFService, this change will only work without side effects when the (already merged) change to do a stream reset inside the loop is present:
https://github.com/vivo-project/Vitro/pull/79

# Interested parties
@VIVO-project/vivo-committers
